### PR TITLE
Search across multiple sources and add msstore source

### PR DESCRIPTION
**Pull request by yao-msft**
This is part 2 of #117

## Change
- Added support to search across multiple sources
- Added store source as one of the default sources if experimental feature is enabled
- Fixed a minor bug where removing a source will always restore default sources

## Validation
- Existing tests passed
- Added new tests to test multiple source search
- Manually tested the test msstore source and community source work together

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-cli/pull/527)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/ManifestComparator.cpp [hunk 1]
```diff
@@ -11,29 +11,28 @@ namespace AppInstaller::CLI::Workflow
...
         // Todo: Compare only architecture for now. Need more work and spec.
-        if (Utility::IsApplicableArchitecture(installer1.Arch) < Utility::IsApplicableArchitecture(installer2.Arch))
+        if (Utility::IsApplicableArchitecture(installer1.Arch) > Utility::IsApplicableArchitecture(installer2.Arch))
         {
-            return false;
+            return true;
         }
 
-        return true;
+        return false;
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/ManifestComparator.cpp:19]: >false [](start = 15, length = 5)

Why did you need to invert this?

File: src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h [hunk 2]
```diff
@@ -14,6 +14,17 @@
...
+    using namespace std::string_view_literals;
+
+    constexpr std::string_view s_Source_WingetCommunityDefault_Name = "winget"sv;
+    constexpr std::string_view s_Source_WingetCommunityDefault_Arg = "https://winget.azureedge.net/cache"sv;
+    constexpr std::string_view s_Source_WingetCommunityDefault_Data = "Microsoft.Winget.Source_8wekyb3d8bbwe"sv;
+
+    constexpr std::string_view s_Source_WingetMSStoreDefault_Name = "msstore"sv;
+    // TODO: update to use prod endpoint
+    constexpr std::string_view s_Source_WingetMSStoreDefault_Arg = "https://winget-test.azureedge.net/msstore"sv;
+    constexpr std::string_view s_Source_WingetMSStoreDefault_Data = "Microsoft.Winget.MSStore.Source_8wekyb3d8bbwe"sv;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h:26]: This should not need to be in the header.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 3]
```diff
@@ -190,6 +186,16 @@ namespace AppInstaller::Repository
                 details.Data = s_Source_WingetCommunityDefault_Data;
                 result.emplace_back(std::move(details));
             }
+
+            if (Settings::ExperimentalFeature::IsEnabled(Settings::ExperimentalFeature::Feature::ExperimentalMSStore))
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:190]: >if [](start = 12, length = 2)

Put the block inside the above one so it is more clear that they are part of the same case.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 4]
```diff
@@ -512,12 +518,36 @@ namespace AppInstaller::Repository
...
+                for (auto& source : currentSources)
+                {
+                    AICLI_LOG(Repo, Info, << "Adding to aggregated source: " << source.Name);
+
+                    if (ShouldUpdateBeforeOpen(source))
+                    {
+                        UpdateSourceFromDetails(source, progress);
+                        sourceUpdated = true;
+                    }
+                    aggregatedSource->AddSource(std::move(CreateSourceFromDetails(source, progress)));
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:541]: >std::move [](start = 48, length = 9)

std::move isn't needed on an rvalue

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 5]
```diff
@@ -512,12 +518,36 @@ namespace AppInstaller::Repository
...
+                auto aggregatedSource = std::make_shared<AggregatedSource>();
+
+                bool sourceUpdated = false;
+                for (auto& source : currentSources)
+                {
+                    AICLI_LOG(Repo, Info, << "Adding to aggregated source: " << source.Name);
+
+                    if (ShouldUpdateBeforeOpen(source))
+                    {
+                        UpdateSourceFromDetails(source, progress);
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:538]: >progress [](start = 56, length = 8)

At some point we are going to have to add a context callback to indicate "we are doing this action now".  This loop could potentially have the progress bar fill up 6+ times with just two sources.

File: src/AppInstallerCLICore/Workflows/ManifestComparator.cpp (hunk 1 above)
Comment (yao-msft) [reply src/AppInstallerCLICore/Workflows/ManifestComparator.cpp:19]: I want to make equal case return false.

---
In reply to: [467208009](https://github.com/microsoft/winget-cli/pull/527#discussion_r467208009) [](ancestors = 467208009)

File: src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h (hunk 2 above)
Comment (yao-msft) [reply src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h:26]: The only reason I move them in the header is s_Source_WingetCommunityDefault_Name is used in AggregatedSource.cpp

Do you want me to move them all back to cpp and create a copy in AggregatedSource.cpp Or keep the *_Name constants here and move *_Arg and *_Data back to cpp? 

---
In reply to: [467209878](https://github.com/microsoft/winget-cli/pull/527#discussion_r467209878) [](ancestors = 467209878)

File: src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySearch.h [hunk 6]
```diff
@@ -111,6 +111,9 @@ namespace AppInstaller::Repository
         // The highest order field on which the application matched the search.
         ApplicationMatchFilter MatchCriteria;
 
+        // The name of the source where the result is from. Used in aggregated source scenario.
+        std::string SourceName = {};
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySearch.h:115]: >std::string SourceName = {}; [](start = 8, length = 28)

A lot of string duplication with this pattern.  The only solution I see to that is to have this or the IApplication just store a shared_ptr to the source itself.  Not necessarily in need of a change, but maybe some thought.

Comment (yao-msft) [reply src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySearch.h:115]: IApplication has a weak pointer to ISource already, then I thought the SearchResult should better be used without having to hold the ISource unless it needs to fetch manifest later.
I also considered making it string_view and store the list of source names somewhere, but it complicates the interface so this ends up as what I can think of here.


---
In reply to: [467338451](https://github.com/microsoft/winget-cli/pull/527#discussion_r467338451) [](ancestors = 467338451)

File: src/AppInstallerRepositoryCore/AggregatedSource.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            result.Truncated = true;
+            result.Matches.erase(result.Matches.begin() + request.MaximumResults, result.Matches.end());
+        }
+
+        return result;
+    }
+
+    void AggregatedSource::SortResultMatches(std::vector<ResultMatch>& matches)
+    {
+        // TODO: for now just simply prefer winget source.
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/AggregatedSource.cpp:52]: If the highest sort priority is source, then no sorting is required.  The sources should be left in their defined order, and so the results will already be good.

The current source already sorts the returned results, first by MatchType, then by Field.  The only other sort option that I have thought of beyond source first is to merge the results in this same sort priority: { MatchType, Field, Source }.  This should allow the results to better represent the search request.

File: src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerRepositoryCore/Public/AppInstallerRepositorySource.h:26]: Move them back to the cpp and don't make a copy elsewhere because they shouldn't be needed.

---
In reply to: [467338302](https://github.com/Microsoft/pkg/pull/527#discussion_r467338302) [](ancestors = 467338302,467209878)

File: src/AppInstallerRepositoryCore/RepositorySource.cpp (hunk 5 above)
Comment (yao-msft) [reply src/AppInstallerRepositoryCore/RepositorySource.cpp:538]: Added todo

---
In reply to: [467334631](https://github.com/microsoft/winget-cli/pull/527#discussion_r467334631) [](ancestors = 467334631)

File: src/AppInstallerRepositoryCore/AggregatedSource.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            // The comparator compares the ResultMatch by MatchType first, then Field in a predefined order.
+            bool operator() (
+                const ResultMatch& match1,
+                const ResultMatch& match2)
+            {
+                std::vector<MatchType> MatchTypeOrder =
+                { MatchType::Exact, MatchType::CaseInsensitive, MatchType::StartsWith, MatchType::Fuzzy,
+                  MatchType::Substring, MatchType::FuzzySubstring, MatchType::Wildcard };
+
+                auto matchTypeOrder1 = std::find(MatchTypeOrder.begin(), MatchTypeOrder.end(), match1.MatchCriteria.Type);
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/AggregatedSource.cpp:63]: >auto matchTypeOrder1 = std::find(MatchTypeOrder.begin(), MatchTypeOrder.end(), match1.MatchCriteria.Type); [](start = 16, length = 106)

This is a very expensive comparator.  I would rather see a comment placed on these enums to state that they need to be in order of preference, then simply compare the integral values of the enums.

File: src/AppInstallerRepositoryCore/AggregatedSource.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        struct ResultMatchComparator
+        {
+            // The comparator compares the ResultMatch by MatchType first, then Field in a predefined order.
+            bool operator() (
+                const ResultMatch& match1,
+                const ResultMatch& match2)
+            {
+                if (match1.MatchCriteria.Type != match2.MatchCriteria.Type)
+                {
+                    return (static_cast<int>(match2.MatchCriteria.Type) - static_cast<int>(match1.MatchCriteria.Type)) > 0;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/AggregatedSource.cpp:61]: >return (static_cast<int>(match2.MatchCriteria.Type) - static_cast<int>(match1.MatchCriteria.Type)) > 0; [](start = 20, length = 103)

```
return ToIntegral(match2.MatchCriteria.Type) > ToIntegral(match1.MatchCriteria.Type);
```
There is no need to do math on them, just direct compare.  The language utilities header has ToIntegral to convert the enum to its actual integral type.
